src/wp-includes/block-editor.php

```diff
@@ -665,6 +665,8 @@ function get_block_editor_settings( array $custom_settings, $block_editor_contex
 		}
 	}
 
+	$editor_settings['canUpdateBlockBindings'] = current_user_can( 'edit_block_binding', $block_editor_context );
```

Comment:
Why is this a new setting and not something you access through `coreData` `canUser` or something?

Reply:
The main reason we started using a new setting was to allow users to disable this behavior with a PHP filter if wanted. And it seems there was a precedence for this with `canLockBlocks`: [link](https://make.wordpress.org/core/2022/05/05/block-locking-settings-in-wordpress-6-0/).

It's true that user capabilities can also be modified through a filter, so if we are able to access `edit_block_binding` in the editor somehow, we wouldn't need an editor setting.

I couldn't find a way to do so, but I must say I'm not familiar with user capabilities.

I'm happy to explore other paths. Is it possible to access these user capabilities? From what I understood, [canUser](https://developer.wordpress.org/block-editor/reference-guides/data/data-core/#canuser) checks whether the user can perform the given action on the given REST resource. But that's not exactly what we need here.

Another possibility to replicate the [capabilities mapping in JS](https://github.com/WordPress/wordpress-develop/pull/7258/files#diff-7b99cb99a8105f5254c9c282a8cb9eed9d9b48f2d974544ce6919fe124bd0427R804-R829), but I am not sure it makes sense and I believe users wouldn't be able to override this behavior easily.

Any thoughts are more than welcome.

Reply:
There are a lot of precedents for block editor settings indeed. But IMO, we should start moving away from them like I explained on my recent blog posts about Gutenberg practices. 

In this case, it's a capability check and we should just have a unique way to fetch capabilities from the client. I think @Mamaduka might know better whether that's already possible or not.

Reply:
For clarity, I'm not saying that users shouldn't override the capability in php, I'm saying that the frontend should access the capability through the capability APIs and not block editor settings.

Reply:
> I'm saying that the frontend should access the capability through the capability APIs and not block editor settings.

If we can make it work this way, I agree it seems the best path forward. I will take another look, but I couldn't find a way to do that, at least with the existing APIs.

Reply:
Block locking was unique because it was not tied to any entity; it was just a general option for blocks, and there was no alternative way of doing it.

I guess the main question here is whether we want to introduce a separate setting that extenders can use to turn block blinding editing on/off or just use capability checks.

P.S. The `wp.data.select( 'core' ).getUser( 1 );` will return all capabilities associated with the user, but I think it should only used when the `canUser` selector can't be applied to the case.

Reply:
> I guess the main question here is whether we want to introduce a separate setting that extenders can use to turn block blinding editing on/off or just use capability checks.

If it is possible to use just capability checks, I think that would suffice.

> P.S. The wp.data.select( 'core' ).getUser( 1 ); will return all capabilities associated with the user, but I think it should only used when the canUser selector can't be applied to the case.

I don't think in this case `canUser` be applied, but I might be mistaken. I tried to use `getCurrentUser` in my tests, but I saw it didn't include the capabilities. I can try with `getUser`.

Reply:
The `getCurrentUser` uses a different endpoint (`/wp/v2/users/me`), which is why it gets different data.

Reply:
From my tests, it seems it is only possible to access the primitive capabilities and not the meta ones. Do you know if it is possible to get them somehow?

Reply:
Not sure. Maybe @TimothyBJacobs or @spacedmonkey could point us in the right direction.

Another options is to use [action links](https://github.com/WordPress/wordpress-develop/blob/2952e563aae229fe497e723996cc5a58b92e1401/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2224), thought I think @youknowriad doesn't like them much 😄

Reply:
I'm not sure about implementation details (action links or not) but I think we should have a selector (`canUser` or another one) to retrieve global capabilities too. maybe like we have a "site" entity, these capabilities can be seen as a site capabilities or something like that.

Reply:
> Block locking was unique because it was not tied to any entity; it was just a general option for blocks, and there was no alternative way of doing it.

To be honest, it doesn't feel unique, it seems like the exact same use case as the current one.

Reply:
> To be honest, it doesn't feel unique, it seems like the exact same use case as the current one.

Probably not, but at the time, shoving everything into block settings seemed like a right decision 😅

Reply:
In PHP we check capabilities using `user_can()`, not `! empty( $user->allcaps['my_cap'] )`. The reason being that `map_meta_cap` ends up being called, as well as other filters. So just checking for the existence of a capability is not enough. For that same reason, we should rarely be looking at the `capabilities` object in the Users REST API, since it's just `allcaps`.

Instead, that's why we should do things like checking the `Allow` header or using "action links" since those all run the actual `user_can()` checks with the appropriate context.

This is an interesting case. As far as I can tell, this only effects Block Editor UI, but there is no equivalent `current_user_can` check being made in the backend to actually _guard_ against someone using this feature.

That indicates to me that this isn't a permission check, but a user preference / editor setting. So I would recommend against using a capability for this purpose. Instead, this looks like something that should be changed by extenders using a filter.

Reply:
> That indicates to me that this isn't a permission check, but a user preference / editor setting. So I would recommend against using a capability for this purpose. Instead, this looks like something that should be changed by extenders using a filter.

In terms of REST API, where would you put such settings?

Reply:
We do have a preferences endpoint/API but these preferences are for things that the user can change. AFAIK, both this setting and the "canBlockLock" setting are not something that the user can change, so they're not really user preferences.

Reply:
> I think it's probably too late in the cycle for that, so this should probably be another one off in wp-block-editor/v1/settings?

If I'm not wrong, that endpoint has never made it into Core.

> I think in an ideal world we'd build up the preferences API to have an equivalent PHP API that could be then used by extenders to filter and force certain preferences on or off.

I'm personally convinced this is not user preferences though. I think we need a consistent global settings endpoint in Core. I know we used the "root" endpoint for that sometimes. Maybe we should just continue using that

Reply:
I guess the difference with the global settings is that this endpoint is user specific. So maybe a "user settings" endpoint.

Reply:
Writing that, I wonder if that's just the user endpoint though :P

Reply:
> If I'm not wrong, that endpoint has never made it into Core.

Yeah looks like it's still GB only. We should probably merge that, I can't remember any specific reasons why we held off on it.

> I'm personally convinced this is not user preferences though. 

I'm not too fussed if it's a "preference" or a "setting". But I do think it's not a capability.

>  I think we need a consistent global settings endpoint in Core. 

Along the lines of `wp-block-editor/v1/settings`?

> I know we used the "root" endpoint for that sometimes. Maybe we should just continue using that

I think this is less of an index type of thing, since it is so spceific to Block Editor UI, and not the general capabiliites of the site.

> I guess the difference with the global settings is that this endpoint is user specific. So maybe a "user settings" endpoint. Writing that, I wonder if that's just the user endpoint though :P

I don't think it would be odd, IMO, for `wp-block-editor/v1/settings` to take into account the user when building the settings. It's already something that I imagine extenders are using when filtering.

Reply:
> I don't think it would be odd, IMO, for wp-block-editor/v1/settings to take into account the user when building the settings. It's already something that I imagine extenders are using when filtering.

Just to be clear, I think a block editor settings endpoint would work here. But I wonder if it's a bit shortsighted in the sense that we'll have a need for similar settings (user settings) outside the block editor. (For instance, a setting to prevent users from creating custom WP-Admin views)...

Reply:
Yeah we should start exploring that in more detail.